receiver/otlpreceiver/otlp.go

```diff
@@ -17,255 +17,140 @@ package otlpreceiver
 import (
 	"context"
 	"errors"
-	"fmt"
 	"net"
 	"net/http"
 	"sync"
 
 	gatewayruntime "github.com/grpc-ecosystem/grpc-gateway/runtime"
-	"github.com/rs/cors"
-	"github.com/soheilhy/cmux"
 	"google.golang.org/grpc"
 
 	"go.opentelemetry.io/collector/component"
-	"go.opentelemetry.io/collector/component/componenterror"
 	"go.opentelemetry.io/collector/consumer"
 	collectormetrics "go.opentelemetry.io/collector/internal/data/opentelemetry-proto-gen/collector/metrics/v1"
 	collectortrace "go.opentelemetry.io/collector/internal/data/opentelemetry-proto-gen/collector/trace/v1"
-	"go.opentelemetry.io/collector/observability"
 	"go.opentelemetry.io/collector/receiver/otlpreceiver/metrics"
 	"go.opentelemetry.io/collector/receiver/otlpreceiver/trace"
 )
 
 // Receiver is the type that exposes Trace and Metrics reception.
 type Receiver struct {
-	mu                sync.Mutex
-	ln                net.Listener
-	serverGRPC        *grpc.Server
-	serverHTTP        *http.Server
-	gatewayMux        *gatewayruntime.ServeMux
-	corsOrigins       []string
-	grpcServerOptions []grpc.ServerOption
+	cfg        *Config
+	serverGRPC *grpc.Server
+	gatewayMux *gatewayruntime.ServeMux
+	serverHTTP *http.Server
 
 	traceReceiver   *trace.Receiver
 	metricsReceiver *metrics.Receiver
 
-	traceConsumer   consumer.TraceConsumer
-	metricsConsumer consumer.MetricsConsumer
-
-	stopOnce                 sync.Once
-	startServerOnce          sync.Once
-	startTraceReceiverOnce   sync.Once
-	startMetricsReceiverOnce sync.Once
-
-	instanceName string
+	stopOnce        sync.Once
+	startServerOnce sync.Once
 }
 
 // New just creates the OpenTelemetry receiver services. It is the caller's
 // responsibility to invoke the respective Start*Reception methods as well
 // as the various Stop*Reception methods to end it.
-func New(
-	instanceName string,
-	transport string,
-	addr string,
-	tc consumer.TraceConsumer,
-	mc consumer.MetricsConsumer,
-	opts ...Option,
-) (*Receiver, error) {
-	ln, err := net.Listen(transport, addr)
-	if err != nil {
-		return nil, fmt.Errorf("failed to bind to address %q: %v", addr, err)
-	}
-
+func New(cfg *Config) (*Receiver, error) {
 	r := &Receiver{
-		ln:          ln,
-		corsOrigins: []string{}, // Disable CORS by default.
-		gatewayMux: gatewayruntime.NewServeMux(
-			gatewayruntime.WithMarshalerOption("application/x-protobuf", &xProtobufMarshaler{}),
-		),
+		cfg: cfg,
 	}
-
-	for _, opt := range opts {
-		opt.withReceiver(r)
+	if cfg.GRPC != nil {
+		opts, err := cfg.GRPC.ToServerOption()
+		if err != nil {
+			return nil, err
+		}
+		opts = append(opts, grpc.ReadBufferSize(512*1024))
```

Comment:
Rebase and use the [new config option](https://github.com/open-telemetry/opentelemetry-collector/blob/7ef6c533d4079e9d640bb7d7ac7b61e999984d4e/config/configgrpc/configgrpc.go#L76) for buffer size. :-)

Reply:
Done.